Adjust minor BeatmapSetOverlay details to better match osu-web

thewildtree opened the pull request:
Follows #7724, aims to bring the `BeatmapSetOverlay` even closer to its osu-web counterpart. These changes are mostly small details, but they definitely can affect the UX.

A few things worth noting:
- the scoreboard font size change from mentioned PR has been reversed - while 12 technically matches osu-web, scaling in lazer is a bit different, making 14 bring a more familiar look here
- scoreboard had wrong order of columns just like `TopScoreStatisticsSection` (as mentioned in #7725), I applied a similar solution, but I'm not sure if it's the proper one here. maybe instead of this we should sort them before they get passed to any UI components?
- added some checks to hide ratings / show placeholder instead of success rate if beatmap has no leaderboard, but I'm not sure if its done the right way
- wanted to add the green colour to the combo if it was the max possible, same as web has, but currently I believe there's no way to check beatmaps maxcombo in this scope

![image](https://user-images.githubusercontent.com/27722894/73866388-1033e080-4845-11ea-8334-caf35e9beb8f.png)

![image](https://user-images.githubusercontent.com/27722894/73866435-23df4700-4845-11ea-8026-f25e7c7b0e64.png)

thewildtree committed:
Simplify GradientLine and fix colour changing

thewildtree committed:
Recolour LeaderboardScopeSelector

thewildtree committed:
Adjust title / artist font weight

thewildtree committed:
Adjust header gradient colours

thewildtree committed:
Adjust PreviewButton alpha and animation

thewildtree committed:
Use alpha instead of colour opacity

thewildtree committed:
Recolour SuccessRate background

thewildtree committed:
Add necessary variable

thewildtree committed:
Adjust online status pill font and padding

thewildtree committed:
Adjust header padding

thewildtree committed:
Rename variable

thewildtree committed:
Recolour DetailBox

thewildtree committed:
Hide ratings if beatmap has no leaderboard

thewildtree committed:
Recolour PreviewButton

thewildtree committed:
Show placeholder instead of success rate when beatmap unranked

thewildtree committed:
Adjust TopScoreUserSection font and spacing

thewildtree committed:
Change scoreboard text size

thewildtree committed:
Enforce correct column order in ScoreTable

thewildtree committed:
Adjust ScoreTable spacing

thewildtree committed:
Adjust corner radius

thewildtree committed:
Adjust font once again for readibility

thewildtree committed:
Add spacing to match osu-web

Note: due to osu-web using flex to even out the spacing and me not being able to implement the same behaviour here, I added a static margin to separate the title from the diffname above. This looks better than the previous state in most cases, the only scenario where this differs somehow visibly from web is on mapsets with large numbers of difficulties.

thewildtree committed:
Fix tests

thewildtree committed:
Remove space

thewildtree committed:
Convert field to local variable

bdach commented:
Please use `.\InspectCode.ps1` to find stylistic issues instead of relying on CI to spot them for you.

thewildtree committed:
Remove using directives

thewildtree committed:
Remove using directive

TheWildTree commented:
Sorry, I couldn't run InspectCode because of some issues with PowerShell on my PC. Will start doing that as soon as I solve them. Also sorry for the CI-fix commit spam, I'm no longer at my PC and github web edit only allows one file per commit.

thewildtree committed:
Add optional decimal place

thewildtree committed:
Recolour ranked status pill

thewildtree committed:
Match osu-web display accuracy

Decided to change this only locally instead of modifying FormatAccuracy which would affect everywhere else in the game as well.

thewildtree committed:
Match osu-web button description

thewildtree committed:
Add alwaysShowDecimals param to FormatAccuracy

This allows us to specify whether we want it to show decimal places if accuracy is 100%.

thewildtree committed:
Change accuracy formatting method

thewildtree committed:
Adjust loading container corner radius

TheWildTree commented:
Fixed few more details I found, I'm gonna leave the `AdvancedStats` bars background like it is even though it doesn't match osu-web since that part is reused elsewhere and that's a very small detail anyway.

Other than that, no more changes planned.

bdach commented on osu.Game/Utils/FormatUtils.cs:
```diff
@@ -7,18 +7,18 @@ public static class FormatUtils
+        public static string FormatAccuracy(this double accuracy, bool alwaysShowDecimals = false) => accuracy == 1 && !alwaysShowDecimals ? "100%" : $"{accuracy:0.00%}";
 
         /// <summary>
         /// Turns the provided accuracy into a percentage with 2 decimal places.
```
Kind of no longer true. I'm overall not too happy about this `alwaysShowDecimals` flag. Why `0.00%` but not `100.00%`? At this point if we're centralising we should centralise, potential discrepancies with web be damned.

TheWildTree replied:
If I understand correctly, this wasn't true beforehand either - `100%` never had decimals shown before this change, so, if anything, it brings that closer to truth if the flag is set to `true`.

> Why `0.00%` but not `100.00%`? 

Good question to ask the original author of this method.

bdach replied:
Honestly feels like wasting time going back and forth on significant digits. Should pick an opinionated choice and stick with it.

TheWildTree replied:
Agreed, I also think we should just stick to 2 decimal places everywhere, as it'ss the most consistent option.

bdach commented on osu.Game/Overlays/BeatmapSet/Scores/ScoreTable.cs:
```diff
@@ -65,6 +66,10 @@ public IReadOnlyList<ScoreInfo> Scores
 
+                // Ensure correct column order
+                foreach (ScoreInfo score in value)
+                    score.Statistics = score.Statistics.OrderByDescending(pair => pair.Key).ToDictionary(pair => pair.Key, pair => pair.Value);
```
Shouldn't be mutating `ScoreInfo` IMO. Would be best if `ScoreInfo` exposed a read-only ordered view of rank-count pairs.

TheWildTree replied:
That would indeed be a good way to do it. 

I think I'll split this as well as the `FormatAccuracy` changes into separate PRs, since their scope is beyond just the beatmapset overlay.

peppy commented on osu.Game/Overlays/BeatmapSet/Scores/ScoreTable.cs:
```diff
@@ -116,7 +121,7 @@ private Drawable[] createContent(int index, ScoreInfo score)
                 {
                     Margin = new MarginPadding { Right = horizontal_inset },
-                    Text = score.DisplayAccuracy,
+                    Text = score.Accuracy.FormatAccuracy(alwaysShowDecimals: true),
```
this shouldn't be happening. we added the `DisplayAccuracy` string for a reason.

peppy replied:
as mentioned elsewhere, please split these changes into a separate PR for further discussion. it's not clear why this is being done in this PR currently.

thewildtree committed:
Revert accuracy display and column sorting changes

TheWildTree commented:
Removed mentioned parts, will create separate PRs for them later.

smoogipoo commented on osu.Game/Overlays/BeatmapSet/Scores/ScoreTable.cs:
```diff
@@ -190,7 +190,7 @@ private class HeaderText : OsuSpriteText
             {
                 Text = text.ToUpper();
-                Font = OsuFont.GetFont(size: 10, weight: FontWeight.Bold);
+                Font = OsuFont.GetFont(size: 12, weight: FontWeight.Bold);
```
osu-web uses size 10 here.

smoogipoo commented on osu.Game/Overlays/BeatmapSet/Scores/ScoreTable.cs:
```diff
@@ -23,7 +23,7 @@ public class ScoreTable : TableContainer
         private const float horizontal_inset = 20;
         private const float row_height = 25;
-        private const int text_size = 12;
+        private const int text_size = 14;
```
osu-web uses size 12 here

smoogipoo commented on osu.Game/Overlays/BeatmapSet/Info.cs:
```diff
@@ -110,6 +112,14 @@ public Info()
+                                    Origin = Anchor.Centre,
+                                    Alpha = 0,
+                                    Text = "Unranked beatmap",
+                                    Font = OsuFont.GetFont(size: 13)
```
osu-web uses size 12 here

smoogipoo commented on osu.Game/Overlays/BeatmapSet/Scores/TopScoreUserSection.cs:
```diff
@@ -96,13 +96,14 @@ public TopScoreUserSection()
                                 Anchor = Anchor.CentreLeft,
                                 Origin = Anchor.CentreLeft,
-                                Font = OsuFont.GetFont(size: 10, weight: FontWeight.Bold)
+                                Font = OsuFont.GetFont(size: 12)
```
osu-web uses size 10 here

smoogipoo reviewed (changes requested):
In general, always match osu-web sizings

TheWildTree commented:
size 10 is unreadable due to the font rendering issues

in general, if you take a look at the rest of the overlay, most elements use fonts a few px bigger than web and I followed that direction - it seems correct because matching osu-web 1:1 here looks way more off (I suppose due to some scaling behaviour I can't properly explain) and font sizes like 10px are unreadable due to how lazer renders them in overlays.

smoogipoo commented:
Match the font sizes. We can uniformly apply a scale at a later point if needed, inside `OsuFont` itself.

TheWildTree commented:
Then we'd have to change every single overlay element / font since they're mostly sized a tad bigger than osu-web.

smoogipoo commented:
That's correct.

TheWildTree commented:
Alright, I'll try to work on it, I'm not sure how that's gonna end up looking though.

peppy commented:
It will look fine. All metrics should match (font sizes, widths, paddings, margins). Overall adjustments can be applied later on when we decide how large or small we want things to display.

The important part is that everything matches on a numeric and relative basis so it is easy to confirm correctness without manually pixel checking (aka what you've probably been doing until now).

thewildtree committed:
Use osu-web font sizes

TheWildTree commented:
Went back to osu-web font size values. I'll make a separate PR soon which will aim to bring all the dimensions / sizes 1:1 with osu-web.

bdach committed:
Extract method to avoid nested ternaries

bdach committed:
Remove unnecessary coercions

Comparisons to null of nullable numbers are always false.

bdach committed:
Adjust font sizes

bdach commented:
I fixed up a few remaining nits; looks fine otherwise.

TheWildTree commented on osu.Game/Overlays/BeatmapSet/Header.cs:
```diff
@@ -134,7 +134,7 @@ public Header()
                                                     title = new OsuSpriteText
                                                     {
-                                                        Font = OsuFont.GetFont(size: 37, weight: FontWeight.SemiBold, italics: true)
+                                                        Font = OsuFont.GetFont(size: 37.5f, weight: FontWeight.SemiBold, italics: true)
```
If the artist text uses osu-web font size, I think this should too

bdach replied:
Good point, I got confused and looked at web's `line-height` instead of `font-size`. Will fix up in a sec.

thewildtree committed:
Match osu-web font size

bdach reviewed (approved):
Think this PR has run its course - I'd say let's get it in for now. Accuracy with web can be iterated on.

thewildtree committed:
Merge branch 'master' into adjust-beatmap-overlay

smoogipoo committed:
Merge branch 'master' into adjust-beatmap-overlay
